File: Glm/probit_bayes_gibbs.jl
using DelimitedFiles, Distributions, LinearAlgebra
affair = readdlm("affair.csv", ',')
y = Int64.(affair[:,1]); # convert the response to integers
X = affair[:,2:end];
XX = X'*X;
n,  k = size(X);
V0 = 10*diagm(ones(k)); #prior covariance
invV0 = V0\I;
burnin = 500;
nloop = 5000+burnin;
store_beta = zeros(nloop,k);
z = 1.0*y; # initial guess, new float copy
beta = XX\(X'*z);
# compute a few things before the loop
id0 = findall(y .== 0); id1 = findall(y .==1);
n0 = length(id0); n1=n-n0;
V = (invV0 + XX)\I; # posterior covariance
for i in  1:nloop 
    # sample z
    global beta
    global Xb = X*beta;
    for k in id0
        z[k] = rand(Truncated(Normal(Xb[k],1), -Inf,0))
    end
    for k in id1
        z[k] = rand(Truncated(Normal(Xb[k],1), 0, Inf))
    end
    # sample beta
    dbeta = X'*z;
    beta = V*dbeta + cholesky(Hermitian(V)).L*randn(k,1);
    store_beta[i,:] = beta';
end
store_beta = store_beta[burnin+1:end,:]; # discard the burn-in
mean(store_beta,dims=1)

# for Problem 9.4
N = size(store_beta)[1];
store_prob = zeros(N);
x = [1 1 10 1 0 16 1];
for loop=1:N
  store_prob[loop] = cdf.(Normal(0,1),x*store_beta[loop,:])[1]
end
include("../../Julia/KDE/ThetaKDE/ThetaKDE.jl")
using .ThetaKDE
kde(store_prob,plt=true)

File: Glm/probit_bayes_gibbs.m
% probit_bayes_gibbs.m
load('affair.csv');
y = affair(:,1); 
X = affair(:,2:end);
XX = X'*X;
[n k] = size(X);
V0 = 10*eye(k);            % prior covariance
invV0 = V0\speye(k);
burnin = 500;
nloop = 5000+burnin;
store_beta = zeros(nloop,k);
z = y;                     % initial guess
beta = XX\(X'*z);
                           % compute a few things before the loop
id0 = find(y==0); id1 = find(y==1);
n0 = length(id0); n1=n-n0;
V = (invV0 + XX)\speye(k); % posterior covariance
for i = 1:nloop
        % sample y^*
    Xb = X*beta;
    z(id0) = tnormrnd(Xb(id0),ones(n0,1),-inf,0);  
    z(id1) = tnormrnd(Xb(id1),ones(n1,1),0,inf); 
        % sample beta
    dbeta = X'*z;
    beta = V*dbeta + chol(V,'lower')*randn(k,1);
	store_beta(i,:) = beta';
end
store_beta = store_beta(burnin+1:end,:);  % discard the burnin
mean(store_beta)
std(store_beta)